diff --git a/ql/src/java/org/apache/hadoop/hive/ql/io/orc/CompressionKind.java b/ql/src/java/org/apache/hadoop/hive/ql/io/orc/CompressionKind.java
index aaba663fc4..2180d9817f 100644
--- a/ql/src/java/org/apache/hadoop/hive/ql/io/orc/CompressionKind.java
+++ b/ql/src/java/org/apache/hadoop/hive/ql/io/orc/CompressionKind.java
@@ -27,7 +27,8 @@
   NONE(org.apache.orc.CompressionKind.NONE),
   ZLIB(org.apache.orc.CompressionKind.ZLIB),
   SNAPPY(org.apache.orc.CompressionKind.SNAPPY),
-  LZO(org.apache.orc.CompressionKind.LZO);
+  LZO(org.apache.orc.CompressionKind.LZO),
+  QAT(org.apache.orc.CompressionKind.QAT);
 
   CompressionKind(org.apache.orc.CompressionKind underlying) {
     this.underlying = underlying;
